### ISSUE TITLE
Compilation error (clangparser.cpp)

### ISSUE THREAD

**Issue by Lecrapouille** [doxygen/doxygen#6946]
SHA1 2c72e1c70c71966c3453d195c441e1cb69195779
Debian 9.8 amd64 with installed llvm-6 and clang-6 (from apt-get).
```
cd doxygen
mkdir build && cd build
cmake -Duse_libclang:BOOL=ON ..
make
```
Got this error:
```
doxygen/src/clangparser.cpp: In member function ‘void ClangParser::linkIdentifier(CodeOutputInterface&, FileDef*, uint&, uint&, const char*, int)’:
doxygen/src/clangparser.cpp:785:59: error: cannot convert from pointer to base class ‘Definition’ to pointer to derived class ‘MemberDef’ because the base is virtual
       addDocCrossReference(g_currentMemberDef,(MemberDef*)d);
```

Note 01: I added `-Duse_libclang:BOOL=ON` to try to fix the warning:
```
warning: Tag `CLANG_ASSISTED_PARSING' at line 1073 of file `Doxyfile' belongs to an option that was not enabled at compile time.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u", or recompile doxygen with this feature enabled.
warning: Tag `CLANG_OPTIONS' at line 1081 of file `Doxyfile' belongs to an option that was not enabled at compile time.
         To avoid this warning please remove this line from your configuration file or upgrade it using "doxygen -u", or recompile doxygen with this feature enabled.
```
Note 02: without this option, compilation worked fine.

**Comment by Lecrapouille** [comment]
I'm not really sure of what I have done but I added a dynamic_cast and compilation ended. I hope not to have a nullptr soon :)

**Comment by albert-github** [comment]
I think this looks like a correct patch when it is something like: ` dynamic_cast<MemberDef*>(d)`
I think the nullptr won't happen as on a previous code line (779) there is a test against the nullptr.

Regarding the LLVM ERROR, I think best is to create a separate option for it (does not have anything to do with the compilation error) also specifying which compiler you are using and the version of clang you are using.

**Comment by Lecrapouille** [comment]
> I think this looks like a correct patch when it is something like: ` dynamic_cast<MemberDef*>(d)`
> I think the nullptr won't happen as on a previous code line (779) there is a test against the nullptr.

I meant dynamic_cast can return nullptr is there is no relation of inheritance between classes. Because I don't know the code I hope it's ok.

> Regarding the LLVM ERROR, I think best is to create a separate issue for it (does not have anything to do with the compilation error) also specifying which compiler you are using and the version of clang you are using.

You are right that may not be related with this compilation error (but appearing with -Duse_libclang:BOOL=ON). I linked cmake with clang-6, I think doxygen was compiled with g++-6

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #6947, based on change proposed in the the comment.
Added possibility to build doxyparse and doxyapp also with clang (as they use the doxygen library).

**Comment by albert-github** [comment]
Code has been integrated in master on github (please don't close the issue as this will be done at the moment of a release).

**Comment by albert-github** [comment]
@Lecrapouille  the problem with the LLVM_ERROR might be related with #6379

**Comment by Lecrapouille** [comment]
@albert-github ! Thanks I'll try within this week. I'll give you a feedback.

**Comment by Lecrapouille** [comment]
@albert-github Finally testing was faster than I was expecting. Removing `support` in doxygen/src/CMakeLists.txt fix the second bug.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.8.16.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
